Support building putty for windows

matthewbauer opened the pull request:
###### Motivation for this change

Uses the mingw makefile in putty to build for Windows targets.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

matthewbauer committed:
windows: refactor

matthewbauer committed:
mingw: add osvi.patch

This is apparently a typo that has not been fixed.

Also remove unused configureFlags.

matthewbauer committed:
gccCrossMingw2: remove

matthewbauer committed:
putty: build on windows

GrahamcOfBorg commented:
Success on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.44632&attempt_id=461ce7e3-cf6b-47bc-80b2-475b63ed38fe)

Attempted: putty

<details><summary>Partial log (click to expand)</summary><p>

```
shrinking /nix/store/c4bcx591az0aplf1m81ij2d2y9c16pk7-putty-0.70/bin/putty
shrinking /nix/store/c4bcx591az0aplf1m81ij2d2y9c16pk7-putty-0.70/bin/plink
shrinking /nix/store/c4bcx591az0aplf1m81ij2d2y9c16pk7-putty-0.70/bin/pterm
shrinking /nix/store/c4bcx591az0aplf1m81ij2d2y9c16pk7-putty-0.70/bin/puttygen
gzipping man pages under /nix/store/c4bcx591az0aplf1m81ij2d2y9c16pk7-putty-0.70/share/man/
strip is /nix/store/1hi76hr87bd1y1q1qjk0lv8nmcjip1c8-binutils-2.30/bin/strip
stripping (with command strip and flags -S) in /nix/store/c4bcx591az0aplf1m81ij2d2y9c16pk7-putty-0.70/bin
patching script interpreter paths in /nix/store/c4bcx591az0aplf1m81ij2d2y9c16pk7-putty-0.70
checking for references to /build in /nix/store/c4bcx591az0aplf1m81ij2d2y9c16pk7-putty-0.70...
/nix/store/c4bcx591az0aplf1m81ij2d2y9c16pk7-putty-0.70
```
</p></details>

GrahamcOfBorg commented:
Success on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.44632&attempt_id=0e803f0b-4275-49b8-babe-9bef07d38ad1)

Attempted: putty

<details><summary>Partial log (click to expand)</summary><p>

```
shrinking /nix/store/767rqz9lqcriyl53k6j5q4gi6amqnc1y-putty-0.70/bin/puttygen
shrinking /nix/store/767rqz9lqcriyl53k6j5q4gi6amqnc1y-putty-0.70/bin/psftp
shrinking /nix/store/767rqz9lqcriyl53k6j5q4gi6amqnc1y-putty-0.70/bin/pscp
shrinking /nix/store/767rqz9lqcriyl53k6j5q4gi6amqnc1y-putty-0.70/bin/plink
gzipping man pages under /nix/store/767rqz9lqcriyl53k6j5q4gi6amqnc1y-putty-0.70/share/man/
strip is /nix/store/zrs21zqcchgyabjf4xfimncdq16njizc-binutils-2.30/bin/strip
stripping (with command strip and flags -S) in /nix/store/767rqz9lqcriyl53k6j5q4gi6amqnc1y-putty-0.70/bin
patching script interpreter paths in /nix/store/767rqz9lqcriyl53k6j5q4gi6amqnc1y-putty-0.70
checking for references to /build in /nix/store/767rqz9lqcriyl53k6j5q4gi6amqnc1y-putty-0.70...
/nix/store/767rqz9lqcriyl53k6j5q4gi6amqnc1y-putty-0.70
```
</p></details>

GrahamcOfBorg commented:
Success on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.44632&attempt_id=7dc68624-5ee5-4895-b26b-2c8d495ba7e8)

Attempted: pkgsCross.mingw32.putty

<details><summary>Partial log (click to expand)</summary><p>

```
        -lshell32 -luser32
installing
post-installation fixup
DLLPATH='/nix/store/15lkdn4wc30yp0qi7p9zakr5cfcr2i5c-putty-0.70-i686-pc-mingw32/bin:/nix/store/bz5izin8r8h72r3p1vbrgfq90pw9dh67-autoconf-2.69/bin:/nix/store/m6idwq9ysc3fv0y1chrcniwz71ygi5b4-automake-1.16.1/bin:/nix/store/knsignig3d67ylsfy617bnjq36dshhri-halibut-1.2/bin:/nix/store/jri0qzh5f41my6sda8vpj2brn37pjnza-libtool-2.4.6/bin:/nix/store/3175j1w4q33kqkf9x0scg9ixndcgfrcb-gnum4-1.4.18/bin:/nix/store/k15ac8wk8h7r30y6g6z5j6rahc28aanm-perl-5.24.4/bin:/nix/store/9nb3k75y84q4lgjzw52k1sh1i34i7q8j-pkg-config-0.29.2/bin:/nix/store/0gahyl2dnjja8648z17jxf47wlb7vsky-patchelf-0.9/bin:/nix/store/a05f8621chsd3h860v8vfyy0g0wgzy7p-paxctl-0.9/bin:/nix/store/jhs9y3f0gkb11afjpza6a5ppd2h6f06w-i686-pc-mingw32-stage-final-gcc-debug-wrapper-7.3.0/bin:/nix/store/pv9zjzci2jafw36cdb0qml8wsi5hnr2i-i686-pc-mingw32-stage-final-gcc-debug-7.3.0/bin:/nix/store/ssr9dhrp11h4lnr35p05bbqc1ln5004d-coreutils-8.29/bin:/nix/store/ws9x9zlnc52y5qr32rxj6lg7wsywyngf-i686-pc-mingw32-binutils-wrapper-2.30/bin:/nix/store/b3fyjpwhqgpjphcn8dw4v2mzh918glxk-i686-pc-mingw32-binutils-2.30/bin:/nix/store/ssr9dhrp11h4lnr35p05bbqc1ln5004d-coreutils-8.29/bin:/nix/store/ssr9dhrp11h4lnr35p05bbqc1ln5004d-coreutils-8.29/bin:/nix/store/7dvq2phf5520wgn2xqq0hbpq2d5iq82d-findutils-4.6.0/bin:/nix/store/i3vi6bipksggp3y11jf8chv99ac174r5-diffutils-3.6/bin:/nix/store/ms0hpa5af9pgjvkjlac5j07xbs09ssb3-gnused-4.5/bin:/nix/store/hnrdvnnzhrppcxqg14bifnnyy4qy94bf-gnugrep-3.1/bin:/nix/store/khp93ad2sbls28z397kx2k4k7lb786ph-gawk-4.2.1/bin:/nix/store/gncd1vwkz1l16qnvdr09cdvphr5agcz8-gnutar-1.30/bin:/nix/store/m90kmkphajn8pns6a7ypgj284jn6nlwg-gzip-1.9/bin:/nix/store/nm3avgd0n7ipk2xc5s0x5352nylsbhjb-bzip2-1.0.6.0.1-bin/bin:/nix/store/34x9zwhm2c4s1n8rklsgxc878czxnf26-gnumake-4.2.1/bin:/nix/store/6vqgi9d49smsbr2qxra52yhipg0yxf9f-bash-4.4-p23/bin:/nix/store/zr5252g93sb8yjhc7l67x9ggsarx4cma-patch-2.7.6/bin:/nix/store/p4n73z8l1hyyyybsycwbrdry55xlhxvd-xz-5.2.4-bin/bin'
Created 0 DLL link(s) in /nix/store/15lkdn4wc30yp0qi7p9zakr5cfcr2i5c-putty-0.70-i686-pc-mingw32/bin
shrinking RPATHs of ELF executables and libraries in /nix/store/15lkdn4wc30yp0qi7p9zakr5cfcr2i5c-putty-0.70-i686-pc-mingw32
i686-pc-mingw32-strip is /nix/store/b3fyjpwhqgpjphcn8dw4v2mzh918glxk-i686-pc-mingw32-binutils-2.30/bin/i686-pc-mingw32-strip
stripping (with command i686-pc-mingw32-strip and flags -S) in /nix/store/15lkdn4wc30yp0qi7p9zakr5cfcr2i5c-putty-0.70-i686-pc-mingw32/bin
patching script interpreter paths in /nix/store/15lkdn4wc30yp0qi7p9zakr5cfcr2i5c-putty-0.70-i686-pc-mingw32
/nix/store/15lkdn4wc30yp0qi7p9zakr5cfcr2i5c-putty-0.70-i686-pc-mingw32
```
</p></details>

GrahamcOfBorg commented:
Failure on x86_64-darwin [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.44632&attempt_id=886f6b7d-90ac-4fcf-98af-4b359a981f0b)

Attempted: putty

<details><summary>Partial log (click to expand)</summary><p>

```
mv -f $depbase.Tpo $depbase.Po
./../unix/gtkcomm.c:41:10: fatal error: 'gdk/gdkx.h' file not found
#include <gdk/gdkx.h>
         ^~~~~~~~~~~~
1 error generated.
make[1]: *** [Makefile:1517: unix/gtkcomm.o] Error 1
make[1]: Leaving directory '/private/tmp/nix-build-putty-0.70.drv-0/putty-0.70/unix'
make: *** [Makefile:1095: all] Error 2
builder for '/nix/store/iicjqpk0ls3jic3112zf6vspia0phk6m-putty-0.70.drv' failed with exit code 2
error: build of '/nix/store/iicjqpk0ls3jic3112zf6vspia0phk6m-putty-0.70.drv' failed
```
</p></details>

Mic92 commented:
At which point Microsoft will hire Nix people?

zimbatm commented on pkgs/os-specific/windows/mingwrt/default.nix:
```diff
@@ -1,7 +1,15 @@
-{ stdenv, callPackage }:
+{ stdenv, callPackage, lib, fetchurl }:
```
it looks like callPackage is not needed anymore

zimbatm reviewed (approved):
good stuff! I have also the nullsoft installer packaged somewhere to windows packages could be more easily redistributed

matthewbauer commented:
Haha... well putty is still free software! I want to get to the point where all of the Windows apps I use can be built from Nix expressions

Ericson2314 commented:
CC @angerman is the bootstrapping still OK (until we can do the C++ stuff at least?)

Ericson2314 commented on pkgs/os-specific/windows/default.nix:
```diff
@@ -0,0 +1,29 @@
+{ newScope, crossLibcStdenv }: let
```
FWIW darwin's went in `pkgs/top-level/darwin-packages.nix` but used to be in `pkgs/os-specific/darwin/default.nix`. I'm not sure which I like better; I think @LnL7 made the switch?

matthewbauer replied:
I prefer this way just for the shorter file paths. Also top-level is starting to get crowded.

LnL7 replied:
I moved `darwin = { ... };` from all-packages to a separate file, I don't think there was a default.nix. Almost every package set points to a `pkgs/top-level/<foo>-packages.nix` so I think that's better for consistency's sake.

angerman commented on pkgs/os-specific/windows/mingw-w64/common.nix:
```diff
@@ -1,16 +1,11 @@
-  configureFlags = [
-    "--enable-idl"
-    "--enable-secure-api"
-  ];
```
I’m pretty sure we want to al least retain the secure-api. Otherwise this will result in frequent missing symbol issues.

matthewbauer replied:
These weren't being used - only name & src are inherited in the derivation below. We can add them back though.

angerman commented on pkgs/os-specific/windows/mingw-w64/osvi.patch:
```diff
@@ -0,0 +1,11 @@
+   WINBOOL IsPlatformNT() {
+     OSVERSIONINFOA oi = { 0 };
+ 
+-    oi.dwOSVersionInfoSize = sizeof (osvi);
```
Why do we do this? A comment in the source would be nice. Lest the next one reading this will be puzzling over this...

matthewbauer replied:
It's a bug I think. See https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=863064.

angerman reviewed (commented):
Shouldn’t mingwrt be called msvcrt for consistency?

angerman commented:
@zimbatm I’ve got a nsis.nix as well that even builds the crosscompiling version (to run makensis on Linux)

matthewbauer committed:
windows: add more meta

‘platforms.windows’ includes all windows platforms. Should prefer this
to using assertions.

matthewbauer committed:
windows: Cleanups

GrahamcOfBorg commented:
Success on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.44632&attempt_id=7bfeed79-3489-4a74-90b2-83329b97091f)

Attempted: putty

<details><summary>Partial log (click to expand)</summary><p>

```
shrinking /nix/store/8c9j6vvbp35qhav77zgdy5sp7111pp60-putty-0.70/bin/putty
shrinking /nix/store/8c9j6vvbp35qhav77zgdy5sp7111pp60-putty-0.70/bin/plink
shrinking /nix/store/8c9j6vvbp35qhav77zgdy5sp7111pp60-putty-0.70/bin/pterm
shrinking /nix/store/8c9j6vvbp35qhav77zgdy5sp7111pp60-putty-0.70/bin/puttygen
gzipping man pages under /nix/store/8c9j6vvbp35qhav77zgdy5sp7111pp60-putty-0.70/share/man/
strip is /nix/store/gpc2wld1s0c6qzx9326cwn1wcx29xzsj-binutils-2.30/bin/strip
stripping (with command strip and flags -S) in /nix/store/8c9j6vvbp35qhav77zgdy5sp7111pp60-putty-0.70/bin
patching script interpreter paths in /nix/store/8c9j6vvbp35qhav77zgdy5sp7111pp60-putty-0.70
checking for references to /build in /nix/store/8c9j6vvbp35qhav77zgdy5sp7111pp60-putty-0.70...
/nix/store/8c9j6vvbp35qhav77zgdy5sp7111pp60-putty-0.70
```
</p></details>

GrahamcOfBorg commented:
Failure on x86_64-darwin [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.44632&attempt_id=04ffd80c-ef3f-440f-847a-bee2098bc3de)

Attempted: putty

<details><summary>Partial log (click to expand)</summary><p>

```
mv -f $depbase.Tpo $depbase.Po
./../unix/gtkcomm.c:41:10: fatal error: 'gdk/gdkx.h' file not found
#include <gdk/gdkx.h>
         ^~~~~~~~~~~~
1 error generated.
make[1]: *** [Makefile:1517: unix/gtkcomm.o] Error 1
make[1]: Leaving directory '/private/tmp/nix-build-putty-0.70.drv-0/putty-0.70/unix'
make: *** [Makefile:1095: all] Error 2
builder for '/nix/store/yadkjq4qhgagiivjm0g89dc9x0pxhpxm-putty-0.70.drv' failed with exit code 2
error: build of '/nix/store/yadkjq4qhgagiivjm0g89dc9x0pxhpxm-putty-0.70.drv' failed
```
</p></details>

bjornfor commented:
@angerman: Based on [this comment](https://github.com/NixOS/nixpkgs/pull/44632#issuecomment-411595831), would you mind sharing/upstreaming makensis?
